src/error.rs

```diff
@@ -34,12 +35,61 @@ struct SerializeError<'a> {
 }
 
 impl Error {
-    /// Returns a new [`Error`] with the provided message.
+    /// Create a new [`Error`] from the provided source.
     ///
     pub fn new(source: DynError) -> Self {
         Self::internal_server_error(source)
     }
 
+    /// Create a new [`Error`] from the provided [`io::Error`]. The status code
+    /// of the error returned will correspond to `source.kind()`.
+    ///
+    pub fn from_io(source: io::Error) -> Self {
+        match source.kind() {
+            io::ErrorKind::AlreadyExists => {
+                // Implies a resource already exists.
+                Self::conflict(Box::new(source))
+            }
+
+            io::ErrorKind::BrokenPipe
+            | io::ErrorKind::ConnectionReset
+            | io::ErrorKind::ConnectionAborted => {
+                // Signals a broken connection.
+                Self::bad_gateway(Box::new(source))
+            }
+
+            io::ErrorKind::ConnectionRefused => {
+                // Suggests the service is not ready or available.
+                Self::service_unavailable(Box::new(source))
+            }
+
+            io::ErrorKind::InvalidData | io::ErrorKind::InvalidInput => {
+                // Generally indicates a malformed request.
+                Self::bad_request(Box::new(source))
+            }
+
+            io::ErrorKind::NotFound => {
+                // Indicates a missing resource.
+                Self::not_found(Box::new(source))
+            }
+
+            io::ErrorKind::PermissionDenied => {
+                // Implies restricted access.
+                Self::forbidden(Box::new(source))
+            }
+
+            io::ErrorKind::TimedOut => {
+                // Implies an upstream service timeout.
+                Self::gateway_timeout(Box::new(source))
+            }
+
+            _ => {
+                // Any other kind is treated as an internal server error.
+                Self::internal_server_error(Box::new(source))
```

Comment:
I want to be conservative here so that way an opinion of what a status code should be doesn't preclude someone from using this function liberally.

Reply:
Also, feel free to submit a PR if you see something that is missing here. I like the idea of highly opinionated parts of the code being evolved over time by heated debates within the community. As long as were all professional, treat each other with respect, and follow the [CODE_OF_CONDUCT.md](https://github.com/zacharygolba/via/blob/main/CODE_OF_CONDUCT.md), I'm game.